### PR TITLE
Implement EventTarget#when

### PR DESCRIPTION
**Pull request by maraisr**
#### 95dcb0e22a99b1709abd7db68af9863a52ae6ed5
<pre>
Implement EventTarget#when
<a href="https://bugs.webkit.org/show_bug.cgi?id=284432">https://bugs.webkit.org/show_bug.cgi?id=284432</a>

Reviewed by NOBODY (OOPS!).

This change extends the EventTarget interface with a new `when` method that
creates a new Observable. Once the user subscribes to this Observable, we add an
EventListener and forward events to its Subscriber&apos;s next handler.

Spec: &lt;<a href="https://wicg.github.io/observable/#event-target-integration">https://wicg.github.io/observable/#event-target-integration</a>&gt;

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.window-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any.worker-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.window-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-from.any.worker-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any.js: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any.worker-expected.txt: Rebaseline.
* LayoutTests/inspector/model/remote-object-get-properties-expected.txt: Rebaseline.
* LayoutTests/inspector/runtime/getPreview-expected.txt: Rebaseline.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSEventTargetCustom.h:
(WebCore::JSEventTargetWrapper::globalObject const): The code generator
would like to call a `globalObject` on its `castedThis`, which does not
exist currently on our custom header.
* Source/WebCore/dom/EventTarget+Observable.idl:
Exposes a `Observable when(event, options)` IDL.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::when):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/ObservableEventListenerOptions.h: Added.
* Source/WebCore/dom/ObservableEventListenerOptions.idl: Added.
* Source/WebCore/dom/ObservableEventTarget.cpp: Added.
(WebCore::createSubscriberCallbackEventTarget):
* Source/WebCore/dom/ObservableEventTarget.h: Added.
</pre>
https://github.com/WebKit/WebKit/commit/95dcb0e22a99b1709abd7db68af9863a52ae6ed5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22019 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117897 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62107 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113842 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32223 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40123 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84986 "Found 2 new test failures: imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any.html imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any.worker.html (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35676 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114826 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25731 "Found 31 new test failures: http/tests/site-isolation/basic-iframe-render-output.html http/tests/site-isolation/basic-iframe.html http/tests/site-isolation/basic-transparent-iframe.html http/tests/site-isolation/document-access.html http/tests/site-isolation/double-iframe.html http/tests/site-isolation/edge-sampling-commit-root-frame-load.html http/tests/site-isolation/frame-access-after-window-open.html http/tests/site-isolation/frame-index.html http/tests/site-isolation/fullscreen.html http/tests/site-isolation/history/add-iframe-while-changing-document-title.html ... (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100680 "Found 29 new API test failures: TestWebKitAPI.SiteIsolation.ChildBeingNavigatedToNewDomainByParent, TestWebKitAPI.SiteIsolation.FocusOpenedWindow, TestWebKitAPI.SiteIsolation.DragEvents, TestWebKitAPI.SiteIsolation.LoadingCallbacksAndPostMessage, TestWebKitAPI.SiteIsolation.PasteGIF, TestWebKitAPI.SiteIsolation.PropagateMouseEventsToSubframe, TestWebKitAPI.SiteIsolation.UserScript, TestWebKitAPI.SiteIsolation.LoadStringAfterOpen, TestWebKitAPI.SiteIsolation.AdvanceFocusAcrossFrames, TestWebKitAPI.SiteIsolation.OpenedWindowFocusDelegates ... (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65425 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25052 "Found 2 new test failures: imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any.html imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any.worker.html (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18815 "Found 2 new test failures: imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any.html imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any.worker.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61739 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95110 "Found 24 new API test failures: TestWebKitAPI.SiteIsolation.ChildBeingNavigatedToNewDomainByParent, TestWebKitAPI.SiteIsolation.FocusOpenedWindow, TestWebKitAPI.SiteIsolation.LoadingCallbacksAndPostMessage, TestWebKitAPI.SiteIsolation.UserScript, TestWebKitAPI.SiteIsolation.LoadStringAfterOpen, TestWebKitAPI.SiteIsolation.NavigateOpenerToProvisionalNavigationFailure, TestWebKitAPI.SiteIsolation.PostMessageWithMessagePorts, TestWebKitAPI.SiteIsolation.LoadRequestOnOpenerWebView, TestWebKitAPI.SiteIsolation.NavigateOpener, TestWebKitAPI.SiteIsolation.GrandchildIframe ... (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18887 "Found 48 new test failures: http/tests/dom/noopener-window-not-targetable.html http/tests/site-isolation/basic-iframe-render-output.html http/tests/site-isolation/basic-iframe.html http/tests/site-isolation/basic-transparent-iframe.html http/tests/site-isolation/document-access.html http/tests/site-isolation/double-iframe.html http/tests/site-isolation/draw-after-cross-origin-navigation-between-two-remote-frames.html http/tests/site-isolation/draw-after-navigation.html http/tests/site-isolation/draw-with-size-after-same-origin-navigation.html http/tests/site-isolation/edge-sampling-commit-root-frame-load.html ... (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121180 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38906 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28929 "Found 45 new test failures: http/tests/site-isolation/basic-iframe-render-output.html http/tests/site-isolation/basic-iframe.html http/tests/site-isolation/basic-transparent-iframe.html http/tests/site-isolation/document-access.html http/tests/site-isolation/double-iframe.html http/tests/site-isolation/draw-after-cross-origin-navigation-between-two-remote-frames.html http/tests/site-isolation/draw-after-navigation.html http/tests/site-isolation/draw-with-size-after-same-origin-navigation.html http/tests/site-isolation/edge-sampling-commit-root-frame-load.html http/tests/site-isolation/frame-access-after-window-open.html ... (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93862 "Found 2 new test failures: imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any.html imported/w3c/web-platform-tests/dom/observable/tentative/observable-event-target.any.worker.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39287 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96938 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93686 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38866 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16683 "Found 47 new test failures: http/tests/site-isolation/basic-iframe-render-output.html http/tests/site-isolation/basic-iframe.html http/tests/site-isolation/basic-transparent-iframe.html http/tests/site-isolation/document-access.html http/tests/site-isolation/double-iframe.html http/tests/site-isolation/draw-after-cross-origin-navigation-between-two-remote-frames.html http/tests/site-isolation/draw-after-navigation.html http/tests/site-isolation/draw-with-size-after-same-origin-navigation.html http/tests/site-isolation/edge-sampling-commit-root-frame-load.html http/tests/site-isolation/frame-access-after-window-open.html ... (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34917 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38804 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44299 "Found 1 new failure in bindings/js/JSEventTargetCustom.h") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38441 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41766 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40156 "Built successfully") | | |

### CODE REVIEW

File: LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any.js
Comment (maraisr) [inline LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any.js:1]: Upstreamed with https://github.com/web-platform-tests/wpt/pull/49633. Error messages are non-normative, as such `Uncaught Error:` would never assert true in WebKit, given we don't prefix `Uncaught`.

File: Source/WebCore/bindings/js/JSEventTargetCustom.h [hunk 1]
```diff
@@ -47,6 +47,8 @@ class JSEventTargetWrapper {
 
     operator JSC::JSObject&() { ASSERT(m_wrapper); return *m_wrapper; }
 
+    JSDOMGlobalObject* globalObject() const { ASSERT(m_wrapper); return JSC::jsCast<JSDOMGlobalObject*>(m_wrapper->globalObject()); }
```
Comment (maraisr) [inline Source/WebCore/bindings/js/JSEventTargetCustom.h]: The CodeGenerator wanted to call a `*castedThis->globalObject()` which produced a "member not defined" error.

I am unsure if this is "right", but it got the code to compile, and tests to pass. 

My question is, should the `globalObject` be resolved form the `m_wrapped` instead?

File: Source/WebCore/dom/ObservableEventTarget.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        return adoptRef(*new SubscriberCallbackEventTarget(context, eventTarget, eventType, options));
+    }
+
+    CallbackResult<void> handleEvent(Subscriber& subscriber) final
+    {
+        if (subscriber.protectedSignal()->aborted())
+            return { };
+
+        if (RefPtr eventTarget = m_eventTarget.get()) {
+            AddEventListenerOptions addEventListenerOptions = { /* capture */ m_options.capture, /* passive */ m_options.passive, /* once */ false, /* signal */ subscriber.protectedSignal() };
```
Comment (maraisr) [inline Source/WebCore/dom/ObservableEventTarget.cpp]: Yep, wasn't a fan either. But `AddEventListenerOptions` is not a struct where [can't just](https://github.com/WebKit/WebKit/blob/da8e66a4fd1c8a97a7e7ec9b02c07cfb3dbb7c6d/Source/WebCore/dom/AddEventListenerOptions.h#L35) `.capture =`, because the struct inherits another one.

I saw this pattern of doing `/* XYZ */` [elsewhere as well](https://github.com/WebKit/WebKit/blob/da8e66a4fd1c8a97a7e7ec9b02c07cfb3dbb7c6d/Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp#L800).

File: Source/WebCore/bindings/js/JSEventTargetCustom.h (hunk 1 above)
Comment (maraisr) [reply Source/WebCore/bindings/js/JSEventTargetCustom.h]: For sake of visibility; these are the impacts on the generated code:

```cpp
// WebKitBuild/Debug/DerivedSources/WebCore/JSEventTarget.cpp
void JSEventTargetPrototype::finishCreation(VM& vm)
{
    Base::finishCreation(vm);
    reifyStaticProperties(vm, JSEventTarget::info(), JSEventTargetPrototypeTableValues, *this);
    bool hasDisabledRuntimeProperties = false;
    if (!(((globalObject())->inherits<JSDOMWindowBase>() || (globalObject())->inherits<JSWorkerGlobalScopeBase>()) && jsCast<JSDOMGlobalObject*>(globalObject())->scriptExecutionContext()->settingsValues().observableEnabled)) {
        hasDisabledRuntimeProperties = true;
        auto propertyName = Identifier::fromString(vm, "when"_s);
        VM::DeletePropertyModeScope scope(vm, VM::DeletePropertyMode::IgnoreConfigurable);
        DeletePropertySlot slot;
        JSObject::deleteProperty(this, globalObject(), propertyName, slot);
    }
    if (hasDisabledRuntimeProperties && structure()->isDictionary())
        flattenDictionaryObject(vm);
    JSC_TO_STRING_TAG_WITHOUT_TRANSITION();
}

static inline JSC::EncodedJSValue jsEventTargetPrototypeFunction_whenBody(JSC::JSGlobalObject* lexicalGlobalObject, JSC::CallFrame* callFrame, typename IDLOperation<JSEventTarget>::ClassParameter castedThis)
{
    SUPPRESS_UNCOUNTED_LOCAL auto& vm = JSC::getVM(lexicalGlobalObject);
    auto throwScope = DECLARE_THROW_SCOPE(vm);
    UNUSED_PARAM(throwScope);
    UNUSED_PARAM(callFrame);
    SUPPRESS_UNCOUNTED_LOCAL auto& impl = castedThis->wrapped();
    if (UNLIKELY(callFrame->argumentCount() < 1))
        return throwVMError(lexicalGlobalObject, throwScope, createNotEnoughArgumentsError(lexicalGlobalObject));
    auto* context = jsCast<JSDOMGlobalObject*>(lexicalGlobalObject)->scriptExecutionContext();
    if (UNLIKELY(!context))
        return JSValue::encode(jsUndefined());
    EnsureStillAliveScope argument0 = callFrame->uncheckedArgument(0);
    auto eventTypeConversionResult = convert<IDLAtomStringAdaptor<IDLDOMString>>(*lexicalGlobalObject, argument0.value());
    if (UNLIKELY(eventTypeConversionResult.hasException(throwScope)))
       return encodedJSValue();
    EnsureStillAliveScope argument1 = callFrame->argument(1);
    auto optionsConversionResult = convert<IDLDictionary<ObservableEventListenerOptions>>(*lexicalGlobalObject, argument1.value());
    if (UNLIKELY(optionsConversionResult.hasException(throwScope)))
       return encodedJSValue();
// NOTE: 🚨 IT WAS THIS *castedThis->globalObject() THAT WAS CAUSING COMPILER ERRORS
    RELEASE_AND_RETURN(throwScope, JSValue::encode(toJS<IDLInterface<Observable>>(*lexicalGlobalObject, *castedThis->globalObject(), throwScope, WebCore::Observable::when(impl, *context, eventTypeConversionResult.releaseReturnValue(), optionsConversionResult.releaseReturnValue()))));
}
```

File: Source/WebCore/dom/ObservableEventTarget.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
```
Comment (rniwa) [inline Source/WebCore/dom/ObservableEventTarget.cpp]: I don't think it's useful to have comments like this for named variables. It's useful for boolean since true/false won't tell us what that is on its own so I'd keep /* once */ and then remove all other comments.

Comment (maraisr) [reply Source/WebCore/dom/ObservableEventTarget.cpp]: Yeah makes sense. Think at the time this was because the arguments aren't named, so it's hard to follow what the arguments are. 

In any case! Thanks for the look, i'll swap them over.
